src/main/java/de/dennisguse/opentracks/util/ExceptionHandler.java

```diff
@@ -0,0 +1,90 @@
+package de.dennisguse.opentracks.util;
+
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+
+import androidx.annotation.NonNull;
+
+import java.util.Arrays;
+
+import de.dennisguse.opentracks.BuildConfig;
+import de.dennisguse.opentracks.ShowErrorActivity;
+
+public class ExceptionHandler implements Thread.UncaughtExceptionHandler {
+
+    private static final int EXCEPTION_FORMAT_MAX_RECURSIVITY = 10;
+
+    private final Context context;
+    private final Thread.UncaughtExceptionHandler defaultExceptionHandler;
+
+    public ExceptionHandler(final Context context, final Thread.UncaughtExceptionHandler defaultExceptionHandler) {
+        this.context = context;
+        this.defaultExceptionHandler = defaultExceptionHandler;
+    }
+
+    @Override
+    public void uncaughtException(@NonNull final Thread thread, @NonNull final Throwable exception) {
+        try {
+            String errorReport = generateErrorReport(formatException(thread, exception));
+            Intent intent = new Intent(context, ShowErrorActivity.class);
+            intent.putExtra(ShowErrorActivity.EXTRA_ERROR_TEXT, errorReport);
+            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            context.startActivity(intent);
+            // Pass exception to OS for graceful handling - OS will report it via ADB
+            // and close all activities and services.
+            defaultExceptionHandler.uncaughtException(thread, exception);
+        } catch (final Exception fatalException) {
+            // do not recurse into custom handler if exception is thrown during
+            // exception handling. Pass this ultimate fatal exception to OS
+            defaultExceptionHandler.uncaughtException(thread, fatalException);
+        }
+    }
+
+    private String formatException(final Thread thread, final Throwable exception) {
+        return formatExceptionRecursive(thread, exception, 0);
+    }
+
+    private String formatExceptionRecursive(final Thread thread, final Throwable exception, final int count) {
+        if (count > EXCEPTION_FORMAT_MAX_RECURSIVITY) {
+            return "Max number of recursive exception causes exceeded!";
+        }
+        // print exception
+        StringBuilder stringBuilder = new StringBuilder();
+        StackTraceElement[] stackTrace = exception.getStackTrace();
+        stringBuilder.append(String.format("Exception in thread \"%s\" %s\n", thread.getName(), exception));
+        // print available stacktrace
+        Arrays.stream(stackTrace).forEach(element -> stringBuilder.append("    at ").append(element).append("\n"));
```

Comment:
Is there no cooler way than manually formatting it (again)?
Like
```
Writer writer = new StringWriter();
exception.printStackTrace(new PrintWriter(writer));
String s = writer.toString();
```

Reply:
Thanks, that simplifies a lot and also takes care of the `Cause`, so no recursion necessary to handle on our own.